### ISSUE TITLE
LC3 decoding is slow

### ISSUE THREAD

**Issue by sam131208** [google/liblc3#24]
Are there plans to improve fixed-point computing? 
     In the decoding comparison test, LC3 is much slower than opus custom's decoding speed. The test samples are 16000 sample rate, 32000 bits rate, frame 10ms, Opus Custom decoding time is 6ms, LC3 is 18~19ms.
    The test platform is a 48MHz Cortex-M0+.

**Comment by asoulier** [comment]
Thanks for your test.

Sure, all the samples processing is made in floating point. If you don't have FPU, the performance is in fact catastrophic.
The performance of floating point processing on common CPU w/FPU, is generally much better than fixed point processing, as you don't have to handle the multiples normalization shifting / rounding.

So you mention :
- ~90 MCPS for LC3 decoding
- ~29 MCPS for Opus decoding

As a comparison, I can share some results I have running Corex M33, with 10 ms frame at 32 kbps :
- LC3 Encoding: 12.1 MCPS   Decoding : 5.1 MCPS
- Opus Encoding : 27.3 MCPS   Decoding : 19.2 MCPS
So, it should run in this configuration more than 2x faster than Opus.

A porting to fixed point will give you a boost around 10x less computation needs (will certainly result in < 10 MCPS).
Unfortunately, there is no plan at this time for a such porting.

Can you share your use case, is it for HFP or LE Audio?
Is there a lot of interest for the porting?

**Comment by sam131208** [comment]
@asoulier 
   Thanks for your reply. 
   I plan to develop a low-cost LE Audio that can run on a 48MHz Cortex-M0+ MCU. Unfortunately, most of these MCUs do not have FPUs.
   I tested the floating-point performance of the opus custom and it took 18.4ms, slightly better than LC3.
   I noticed that your test data actually has some differences between OPUS and OPUS Custom, OPUS consists of Silk and Celt, while OPUS Custom only applies Celt.
  OPUS uses 48kHz by default when encoding, decoding is also 48kHz, and finally resampling to the target frequency. This is not the case with Opus Custom (CELT), which is encoded and decoded at the frequency of interest without resampling.

**Comment by asoulier** [comment]
Sounds good, but definitely needs to run in fixed point.
I have done a quick porting of the MDCT [here](https://github.com/google/liblc3/tree/fixed_point).
Of course will be not enough, but that will be time consuming to port all the code.

About the benchmarks, they are a bit old, and I cannot remember exactly what are the parameters of Opus.
But I think it's probably using "Low latency" application mode (disable SILK I guess), and complexity set to 0.
By the way, CELT and LC3 core algorithms are quite close, looks normal that performance are also somewhat close.

**Comment by sam131208** [comment]
@asoulier 
    I tested the new fixed-point branch and the decoding speed was greatly improved, about 14ms. However, there is a function lc3_ltpf_synthesize(), which sometimes consumes a lot of computing power, increasing the time from 14ms to 36ms.
   Expect you to have new good news。

**Comment by asoulier** [comment]
Yes, it's normal, there is a FIR that is activated when a pitch is detected by the encoder.
On higher bitrates, this filter is always disabled.
This filter needs also to be ported in fixed point.
